swarm/storage/mru/handler.go

```diff
@@ -0,0 +1,490 @@
+	if err != nil {
+		return nil, NewError(ErrNotFound, err.Error())
+	}
+
+	// create the index entry
+	rsrc := &resource{}
+
+	if err := rsrc.ResourceMetadata.binaryGet(chunk.SData); err != nil { // Will fail if this is not really a metadata chunk
+		return nil, err
+	}
+
+	rsrc.rootAddr, rsrc.metaHash = metadataHash(chunk.SData)
+	if !bytes.Equal(rsrc.rootAddr, rootAddr) {
+		return nil, NewError(ErrCorruptData, "Corrupt metadata chunk")
+	}
+	h.set(rootAddr, rsrc)
+	log.Trace("resource index load", "rootkey", rootAddr, "name", rsrc.ResourceMetadata.Name, "starttime", rsrc.ResourceMetadata.StartTime, "frequency", rsrc.ResourceMetadata.Frequency)
+	return rsrc, nil
+}
+
+// update mutable resource index map with specified content
+func (h *Handler) updateIndex(rsrc *resource, chunk *storage.Chunk) (*resource, error) {
+
+	// retrieve metadata from chunk data and check that it matches this mutable resource
+	var r SignedResourceUpdate
+	if err := r.fromChunk(chunk.Addr, chunk.SData); err != nil {
+		return nil, err
+	}
+	log.Trace("resource index update", "name", rsrc.ResourceMetadata.Name, "updatekey", chunk.Addr, "period", r.period, "version", r.version)
+
+	// update our rsrcs entry map
+	rsrc.lastKey = chunk.Addr
+	rsrc.period = r.period
+	rsrc.version = r.version
+	rsrc.updated = time.Now()
+	rsrc.data = make([]byte, len(r.data))
+	rsrc.multihash = r.multihash
+	copy(rsrc.data, r.data)
+	rsrc.Reader = bytes.NewReader(rsrc.data)
+	log.Debug("resource synced", "name", rsrc.ResourceMetadata.Name, "updateAddr", chunk.Addr, "period", rsrc.period, "version", rsrc.version)
+	h.set(chunk.Addr, rsrc)
+	return rsrc, nil
+}
+
+// Update adds an actual data update
+// Uses the Mutable Resource metadata currently loaded in the resources map entry.
+// It is the caller's responsibility to make sure that this data is not stale.
+// Note that a Mutable Resource update cannot span chunks, and thus has a MAX NET LENGTH 4096, INCLUDING update header data and signature. An error will be returned if the total length of the chunk payload will exceed this limit.
+func (h *Handler) Update(ctx context.Context, r *SignedResourceUpdate) (storage.Address, error) {
+	return h.update(ctx, r)
+}
+
+// create and commit an update
+func (h *Handler) update(ctx context.Context, r *SignedResourceUpdate) (updateAddr storage.Address, err error) {
+
+	// we can't update anything without a store
+	if h.chunkStore == nil {
+		return nil, NewError(ErrInit, "Call Handler.SetStore() before updating")
+	}
+
+	rsrc := h.get(r.rootAddr)
+	if rsrc != nil && rsrc.period != 0 && rsrc.version != 0 && // This is the only cheap check we can do for sure
+		rsrc.period == r.period && rsrc.version >= r.version { // without having to lookup update chunks
+
+		return nil, NewError(ErrInvalidValue, "A former update in this period is already known to exist")
+	}
+
+	chunk, err := r.toChunk() // Serialize the update into a chunk. Fails if data is too big
```

Comment:
Leave for a new PR ;-) Maybe this could be more global, like being able to query the swarm node's parameters / network information / node info .... One of those could be MRU max size.

Reply:
well, but the max size will depend on the name? Or is the actual metadata of update chunks constant across resources?

Reply:
The metadata of update chunks is a constant and does not depend of the "name"
The name is part of the root metadata chunk and limited to 255 chars and therefore is stored only once.

The max payload data of a resource update as defined in `update.go` is:
`const maxUpdateDataLength = chunkSize - signatureLength - updateHeaderLength - chunkPrefixLength
`. This is 3954 bytes.

Reply:
Gotcha. Nice round number :)

Reply:
https://github.com/ethersphere/go-ethereum/issues/802

Reply:
You gotta have inexplicably convoluted constants in your code for people to take you serious.